### ISSUE TITLE
Closed longpoll/continuous _changes client connections kept alive on server

### ISSUE THREAD

**Issue by jlami** [apache/couchdb#4536]
## Description

We are having memory issues on our cluster. We have tracked it down to replication requests that have a huge amount of (binary) memory open during the crash. This replication is also handling attachments, so that explains why any one of those requests takes more memory.

When looking at prometheus data we saw that `couchdb_httpd_clients_requesting_changes_total` keeps rising:
![image](https://user-images.githubusercontent.com/8522993/232317605-dc3eedc8-7f21-457b-8924-0be923ef19df.png)
(here you see two crashes of couchdb06)

We know that we might need to investigate why the changes feed is queried this often, but that is made more difficult with this bug present.

This led us to test simple _changes requests with longpoll and high timeouts and see that the prometheus data was indeed showing that the connections are kept open (or at least the erlang process is kept alive).

## Steps to Reproduce

1. Look at http://localhost:5984/_node/_local/_prometheus
2. See that couchdb_httpd_clients_requesting_changes_total is 0
3. Start a request to a local database with something like this: http://localhost:5984/db/_changes?feed=longpoll&timeout=200000&since=now
4. Stop and start that request a couple of times in quick succession
5. (optional)check with netstat that the connections are actually closed
6. Refresh the prometheus page and see the couchdb_httpd_clients_requesting_changes_total rising and only falling after the timeout is over

## Expected Behaviour
The value for couchdb_httpd_clients_requesting_changes_total should be falling immediately at connection termination.

## Your Environment
* CouchDB version used: 3.3.1
* Browser name and version: N/A
* Operating system and version: both Windows and Linux have this issue

**Comment by nickva** [comment]
Thank you for your report @jlami.

With `http://localhost:5984/db/_changes?feed=longpoll&timeout=200000&since=now` or `continuous` the client probably stops during the HTTP protocol phase when the client is waiting on the server to respond. If the client disappears, the server won't find out about the connection state until it tries to respond to the client.

Since there is a `timeout`,  at the end of the timeout interval is when the server will try to send a response and find out the client has closed the connection.  In this case the `timeout` is not `200000` but actually `60000`. The `timeout` value is the minimum value between the parameter for the request and the the [changes_timeout](https://docs.couchdb.org/en/stable/config/http.html#chttpd/changes_timeout) setting as per https://github.com/apache/couchdb/blob/4b8b7ec9f39d1e5910f301283fd6b9051b057d0c/src/couch/src/couch_changes.erl#L400-L401 So the connection will be torn down after 60 seconds.

So one way to shorted the time until the connection is cleared is to lower that timeout either in the request parameter or the config setting.

Another option is to use a heartbeat.  

```
/_changes'?feed=longpoll&heartbeat=5000&since=now'
```

This will ping the connection every 5 second and send a newline. If the client dropped the request, the newline send should detect the connection shutdown and clean up the connection resources on the server.

**Comment by jlami** [comment]
Thanks for the reply @nickva

I think the heartbeat might be a good addition/change to our setup to try and hasten the detection. That hopefully helps the application side.

The other thing is that this might also happen because of continuous replication from couchdb servers. I don't think they have the option to enable heartbeats in the _replicate and _replicator endpoints. Or is there a way to set this up?

Is it not possible to detect a disconnect on the erlang side? I cannot really find where the connection is setup in the erlang source, but can't the process that handles the reading of the request also try and detect `tcp_closed` for a socket? I know this will only detect 'clean' connection closing (FIN packets). But it would at least be something.

**Comment by nickva** [comment]
For replication it's possible to adjust [connection_timeout](https://docs.couchdb.org/en/stable/config/replicator.html#replicator/connection_timeout) setting. That setting value, divided by 3, ends up being sent as the `_changes` feed `timeout` connection parameter. So if that's `30000` msec, it will use `...&timeout=10000` in the request. Shortening that could help a bit perhaps. In the normal operation, when the feed is waiting on db changes, the effect will be that absent any changes, the _changes feed request will end and restart every `timeout` milliseconds.

An apparent, improvement could be to improve couch replicator to use both a heartbeat and a timeout.  It turns out both are not really compatible, and the presence of a heartbeat overrides the timeout parameter, so the larger improvement would be to fix that actually, to allow both to work.

> Is it not possible to detect a disconnect on the erlang side? I cannot really find where the connection is setup in the erlang source, but can't the process that handles the reading of the request also try and detect tcp_closed for a socket? I know this will only detect 'clean' connection closing (FIN packets). But it would at least be something.

That might be possible. It would be something in the HTTP layer mochiweb library we're using, I think. But it would be an extra periodic check or something of that sort to monitor for TCP shutdown or disconnect messages even when we're expecting  server to respond (we're not actually waiting in a socket receive).  That logic might go somewhere in here: https://github.com/mochi/mochiweb/blob/83fd22d620bbefd8779ccb6ea21706bb8a1e0ed2/src/mochiweb_http.erl#L139-L140

**Comment by jlami** [comment]
> For replication it's possible to adjust [connection_timeout](https://docs.couchdb.org/en/stable/config/replicator.html#replicator/connection_timeout) setting. That setting value, divided by 3, ends up being sent as the _changes feed timeout connection parameter. So if that's 30000 msec, it will use ...&timeout=10000 in the request. Shortening that could help a bit perhaps. In the normal operation, when the feed is waiting on db changes, the effect will be that absent any changes, the _changes feed request will end and restart every timeout milliseconds.

Unfurtunately our replication needs quite hight timeouts or it will timeout during normal operations.

> An apparent, improvement could be to improve couch replicator to use both a heartbeat and a timeout. It turns out both are not really compatible, and the presence of a heartbeat overrides the timeout parameter, so the larger improvement would be to fix that actually, to allow both to work.

This might indeed be nice. Heartbeat now overrides the timeout, but they might be good to work together.
But I'm correct in inferring that the replicate does not support heartbeat at this moment right?

> That might be possible. It would be something in the HTTP layer mochiweb library we're using, I think. But it would be an extra periodic check or something of that sort to monitor for TCP shutdown or disconnect messages even when we're expecting server to respond (we're not actually waiting in a socket receive). That logic might go somewhere in here: https://github.com/mochi/mochiweb/blob/83fd22d620bbefd8779ccb6ea21706bb8a1e0ed2/src/mochiweb_http.erl#L139-L140

Thanks, I was just looking at mochiweb. I'll see if I can get a feature request/issue for the tcp_closed during handling of the request body.

**Comment by nickva** [comment]
> But I'm correct in inferring that the replicate does not support heartbeat at this moment right?

Yeah, that's correct. It uses timeouts only. So it periodically ends the continuous requests if there are not changes and restarts it.

> I'll see if I can get a feature request/issue for the tcp_closed during handling of the request body.

Mochiweb is a separate project and Apache CouchDB at least in the case of long running requests is sort of abusing the typical HTTP API request pattern. User makes a request, and the callback instead of responding right away, and thus detecting a closed socket and crashing, hangs for minutes at a time. In other words, it won't be surprising if they don't want to accept this feature request.

**Comment by jlami** [comment]
Thanks for the feedback.

> Mochiweb is a separate project and Apache CouchDB at least in the case of long running requests is sort of abusing the typical HTTP API request pattern.

Would that mean that a fix would be more fitting in CouchDB? Don't mean to push the issue onto CouchDB just thinking about what would be best.

### Semi new issue

Since we started investigating our memory issues and we found this issue, we also see that the inflated changes metric might not only be due to the timeouts. Or at least there is something else fishy going on. I don't know if you want me to open a separate issue for that, but I think it fits here.

Looking at the following graph you can see 2 things:
- Orange is the changes metric
- Green is the amount of sockets reported as allocated minus the sockets in use according to `/proc/net/sockstat`
![image](https://user-images.githubusercontent.com/8522993/233593085-d6fa7b6b-59bd-443d-b2aa-8a13af3dd7ad.png)
![image](https://user-images.githubusercontent.com/8522993/233602403-1e6e6963-c594-42e9-83a6-9a8046ed67a6.png)

The drops in the lines to 0 are restarts of couchdb because of memory issues.

The first graph is of a time with extra replication going on, so inflated changes listeners. The second is with mainly external pouch listeners.

When looking at the first image and an erlang dump from around 14:00 I could see that the changes processes where taking a lot of memory and as such found the original issue. But if this was the only problem we would expect the counts of the changes to eventually drop at lower traffic times. And we never see that. Even if you look at the night time, when there are far fewer requests coming in. It seems that the timeouts are never triggered and CouchDB is hanging somewhere else.

There are some half drops in the changes and sockets around 10:00 in the first and 18:00 and 10:00 in the second image. At that time another node was restarted. So it seems that a bunch of the lingering changes requests are inter-node based.

Another thing I can see is that there seems to be a 10 minute delay between the two graphs, exactly the timeout of our chttpd.changes_timeout config value. But the pouch clients request timeouts of 2 minutes.

I'm going to see if I can get a sense of which processes are in what call graph at the time of such high changes metric and lingering sockets. But maybe you could help me what else I could be looking into?

And one thing you might be able to answer: is it normal to see this many unused sockets? We do have many clients with bad network connections, so that might be causing network issues. But I would expect CouchDB to be able to handle this.

**Comment by nickva** [comment]
> Would that mean that a fix would be more fitting in CouchDB? Don't mean to push the issue onto CouchDB just thinking about what would be best.

Yeah I think so. I tried various options and played with sockets a bit and it seems, unless the socket mode is toggled from passive/active or we try to send something on it, we might not find out it's closed. Getting its info or stats doesn't seem to indicate it's closed even.

The new experimental `socket` network implementation may have something there but I haven't tried it yet: https://www.erlang.org/doc/man/socket.html

One obvious improvement though could be to combine heartbeats and timeouts. So even with a timeout set we still send periodic heartbeats, even if it's just to be able to detect a closed socket.

**Comment by jlami** [comment]
I was also experimenting a bit and I could get it to work with something like this in chttpd:handle_changes_req1

```erlang
        Feed when Feed =:= "continuous"; Feed =:= "longpoll"; Feed =:= "eventsource" ->
        	OldControlling = self(),
        	MochiReq = Req#httpd.mochi_req,
        	Socket = MochiReq:get(socket),
        	couch_log:info("----- socket1 ~p -----", [Socket]),
        	SocketMonitorPid = spawn_link(fun() ->
        		couch_log:info("----- socket2 ~p -----", [Socket]),
        		receive ack -> ok end,
        		inet:setopts(Socket, [{active, once}]),
        		receive
        			ack ->
        				couch_log:info("----- stopping monitor -----", []),
        				inet:setopts(Socket, [{active, false}]),
        				gen_tcp:controlling_process(Socket, OldControlling),
        				OldControlling ! ack;
        			{tcp, _} ->
        				couch_log:info("----- data -----", []),
        				couch_stats:decrement_counter([couchdb, httpd, clients_requesting_changes]),
        				exit(self(), {unexpected_data});
        			{tcp_closed, _} ->
        				couch_log:info("----- closed -----", []),
        				couch_stats:decrement_counter([couchdb, httpd, clients_requesting_changes]),
        				error(tcp_closed_error)
        		end
        	end),
        	couch_log:info("----- spawn ~p -----", [SocketMonitorPid]),
        	gen_tcp:controlling_process(Socket, SocketMonitorPid),
        	SocketMonitorPid ! ack,
            couch_stats:increment_counter([couchdb, httpd, clients_requesting_changes]),
            Acc0 = #cacc{
                feed = list_to_atom(Feed),
                mochi = Req,
                threshold = Max
            },
            try
                fabric:changes(Db, fun changes_callback/2, Acc0, ChangesArgs)
            after
            	SocketMonitorPid ! ack,
            	couch_log:info("----- done with changes -----", []),
            	receive ack-> ok end,
            	couch_log:info("----- socket back -----", []),
                couch_stats:decrement_counter([couchdb, httpd, clients_requesting_changes])
            end;
```

But I'm new at erlang so there might be something easier to read from the socket in parallel. And this also might have a race condition when the socket is done and should get more data, but I think that is not legal for change feeds, but maybe it is for Keep-Alive sockets.

**Comment by nickva** [comment]
Making sure streaming requests processes exit when connections close fix was deployed to 3.3.3: https://github.com/apache/couchdb/pull/4736

This should prevent connections from piling up.
